File: python/taflow/metrics/information_ratio.py
"""Benchmark-relative active-return information ratio metric."""

from __future__ import annotations

from typing import Any

from .._native.metrics import InformationRatio as _Native
from ._input import as_paired_metric_series


class InformationRatio:
    """Compute mean aligned active return divided by tracking error.

    Active return is primary simple return minus benchmark simple return, and
    tracking error is its sample standard deviation with one degree of freedom.
    The result is multiplied by ``sqrt(periods_per_year)`` when
    ``annualized=True`` (the default); set ``annualized=False`` for the
    per-period ratio. The independent oracles are Empyrical Reloaded 0.5.12
    ``excess_sharpe`` and QuantStats 0.0.81 ``information_ratio`` for the
    unannualized definition; TAFlow's annualized variant applies the explicitly
    documented scale.

    At least two usable aligned pairs and nonzero tracking error are required;
    otherwise warm-up or a zero denominator yields ``None``. Missing values are
    omitted pairwise under the default ``nan_policy="omit"`` or rejected by
    ``"raise"``. Infinities and mismatched series lengths are rejected before
    state mutation. Inputs
    may be aligned decimal simple returns, log returns, positive equity levels,
    or non-cumulative period P&L. P&L conversion requires separate positive
    initial capital for primary and benchmark streams. The first equity pair
    establishes baselines and does not increment length. Mutating lifecycle
    methods are fluent; Rust owns conversion and O(1)-memory arithmetic.
    """

    def __init__(self, periods_per_year: float = 252.0, annualized: bool = True, nan_policy: str = "omit") -> None:
        """Initialize an empty configured metric."""
        self._state = _Native(float(periods_per_year), annualized, nan_policy)

    def from_returns(self, returns: Any, benchmark_returns: Any) -> "InformationRatio":
        """Append aligned returns series and return this metric."""
        primary, benchmark = as_paired_metric_series(returns, benchmark_returns)
        self._state.from_returns(primary, benchmark)
        return self

    def from_log_returns(self, log_returns: Any, benchmark_log_returns: Any) -> "InformationRatio":
        """Append aligned log returns series and return this metric."""
        primary, benchmark = as_paired_metric_series(log_returns, benchmark_log_returns)
        self._state.from_log_returns(primary, benchmark)
        return self

    def from_equity(self, equity: Any, benchmark_equity: Any) -> "InformationRatio":
        """Append aligned equity series and return this metric."""
        primary, benchmark = as_paired_metric_series(equity, benchmark_equity)
        self._state.from_equity(primary, benchmark)
        return self

    def from_pnl(self, pnl: Any, benchmark_pnl: Any, initial_capital: float, benchmark_initial_capital: float) -> "InformationRatio":
        """Append aligned period P&L with separate initial capitals."""
        primary, benchmark = as_paired_metric_series(pnl, benchmark_pnl)
        self._state.from_pnl(primary, benchmark, float(initial_capital), float(benchmark_initial_capital))
        return self

    def append(self, value: float, benchmark_value: float) -> "InformationRatio":
        """Append one aligned pair in the selected domain and return this metric."""
        self._state.append(float(value), float(benchmark_value))
        return self

    def extend(self, values: Any, benchmark_values: Any) -> "InformationRatio":
        """Append aligned series after equal-length validation and return this metric."""
        primary, benchmark = as_paired_metric_series(values, benchmark_values)
        self._state.extend(primary, benchmark)
        return self

    @property
    def value(self) -> float | None:
        """Return the ratio, or ``None`` until its denominator is defined."""
        return self._state.value

    def compute(self) -> float | None:
        """Return the current scalar without replaying processed input."""
        return self._state.compute()

    def reset(self) -> "InformationRatio":
        """Clear observations, preserve configuration, and return this metric."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the number of usable aligned return pairs processed by Rust."""
        return len(self._state)


__all__ = ["InformationRatio"]

File: crates/taflow-metrics/src/metrics/information_ratio.rs
use crate::{
    primitives::OnlineMoments, MetricError, MetricInputKind, MetricResult, NanPolicy,
    PairedMetricInputState,
};

/// Mean aligned active return divided by its sample standard deviation.
#[derive(Debug, Clone)]
pub struct InformationRatio {
    input: PairedMetricInputState,
    active_return_moments: OnlineMoments,
    annualization_scale: f64,
}

impl InformationRatio {
    /// Construct an empty state for aligned primary and benchmark input domains.
    pub fn new(
        primary_input_kind: MetricInputKind,
        benchmark_input_kind: MetricInputKind,
        periods_per_year: f64,
        annualized: bool,
        nan_policy: NanPolicy,
    ) -> MetricResult<Self> {
        if !periods_per_year.is_finite() || periods_per_year <= 0.0 {
            return Err(MetricError::InvalidParameter {
                name: "periods_per_year",
                value: periods_per_year.to_string(),
                reason: "must be finite and greater than zero",
            });
        }
        if matches!(
            primary_input_kind,
            MetricInputKind::RawPnl | MetricInputKind::Trades
        ) || matches!(
            benchmark_input_kind,
            MetricInputKind::RawPnl | MetricInputKind::Trades
        ) {
            return Err(MetricError::InvalidParameter {
                name: "input_kind",
                value: format!("{primary_input_kind:?}/{benchmark_input_kind:?}"),
                reason: "information ratio requires returns, log returns, equity, or period P&L with initial equity",
            });
        }

        Ok(Self {
            input: PairedMetricInputState::new(
                primary_input_kind,
                benchmark_input_kind,
                nan_policy,
            )?,
            active_return_moments: OnlineMoments::new(),
            annualization_scale: if annualized {
                periods_per_year.sqrt()
            } else {
                1.0
            },
        })
    }

    /// Append one aligned pair and return the current information ratio.
    pub fn append(&mut self, primary: f64, benchmark: f64) -> MetricResult<Option<f64>> {
        if let Some((primary_return, benchmark_return)) = self.input.append(primary, benchmark)? {
            self.active_return_moments
                .append(primary_return - benchmark_return);
        }
        Ok(self.value())
    }

    /// Append equal-length aligned slices through the same persistent state.
    pub fn extend(&mut self, primary: &[f64], benchmark: &[f64]) -> MetricResult<Option<f64>> {
        self.input
            .extend_slices(primary, benchmark, |primary_return, benchmark_return| {
                self.active_return_moments
                    .append(primary_return - benchmark_return);
            })?;
        Ok(self.value())
    }

    /// Return the ratio, or `None` before two pairs or at zero tracking error.
    pub fn value(&self) -> Option<f64> {
        let tracking_error = self.active_return_moments.standard_deviation(1)?;
        if tracking_error == 0.0 {
            return None;
        }
        Some(self.active_return_moments.mean()? / tracking_error * self.annualization_scale)
    }

    /// Return the current result without replaying prior observations.
    pub fn compute(&self) -> Option<f64> {
        self.value()
    }

    /// Restore fresh-state behavior while preserving input and annualization settings.
    pub fn reset(&mut self) {
        self.input.reset();
        self.active_return_moments.reset();
    }

    /// Return the number of usable aligned return pairs processed.
    pub fn len(&self) -> usize {
        self.input.len()
    }

    /// Return whether no usable aligned return pairs have been processed.
    pub fn is_empty(&self) -> bool {
        self.input.is_empty()
    }
}
